pkg/kubelet/cm/dra/manager.go

```diff
@@ -99,6 +100,13 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 				pod.Name, pod.UID, claimName, resourceClaim.UID)
 		}
 
+		// If no container actually uses the claim, then we don't need
+		// to prepare it.
+		if !claimIsUsedByPod(podClaim, pod) {
+			klog.V(5).InfoS("Skipping unused resource", "claim", claimName, "pod", pod.Name)
+			continue
+		}
+
```

Comment:
Something seems off about the placement of this check in this function.

I _believe_ it still works, but it may unnecessarily delay the underyling `NodeUnprepareResources()` in certain scenarios.

The scenarios to consider are:
1. This is the only pod to reference the claim
2. This is the first pod to reference the claim (and more than one eventually reference it)
3. This is not the first pod to reference the claim

In scenario 1, `NodePrepareResources()` will never be called (because of the new check), and `NodeUnprepareResources()` will also never be called (because a claimInfo object for it will never be created).

In scenario 2, `NodePrepareResources()` will not be called when the first pod is started (because of the new check), a reference to the pod will also *not* be added to the `claimInfo` object. Future calls to `PrepareResources()` will have their pods added to the `claimInfo` as appropriate and future calls to `UnprepareResources()` will only trigger an underlying `NodeUnprepareResources()` once the last pod reference in the claimInfo has called it.

In scenario 3, `NodePrepareResources()` will have already been trigged by a previous pod, and since our new check for `!claimIsUsedByPod()` is further down in the code, a reference to this pod *will* be added to the `claimInfo` object. Future calls to `UnprepareResources()` will only trigger an underlying `NodeUnprepareResources()` once the last pod reference in the claimInfo has called it (even if the only pod we are waiting for is one that would fail the `!claimIsUsedByPod()` check.

Reply:
I placed it here because I wanted to do the "is reserved for" check also when the claim is not actually used. We still want to do proper scheduling despite that, which means the claim must be reserved for the pod to run.

What I missed is the 
```
			claimInfo.addPodReference(pod.UID)
			continue
```
above when the claim is already prepared.

I think that's actually another bug in the code: if the claim was prepared for some other pod and then this pod here does't have it reserved, it's allowed to run by kubelet because the "is reserved for" check gets skipped.

Reply:
I fixed this by moving the "is already prepared" check.

Reply:
Seems reasonable. I believe the initial thinking was to exit early to avoid a call to the API server with `ResourceV1alpha2().ResourceClaims(pod.Namespace).Get()`, but I agree that checking `reservedFor` is more important than this minor optimization.